File: backend/app/schemas/inventory.py
"""Inventory and Ticket Schemas"""
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List


class InventoryItem(BaseModel):
    """Single row from the inventory table (current on-hand stock)."""

    id: int
    product_id: int
    category: Optional[str] = None
    article_number: str
    serial_number: Optional[str] = None
    quantity_available: int
    tracking_type: str  # SERIALIZED | BULK
    status: str
    received_by: str
    received_at: Optional[datetime] = None
    purchase_order_id: Optional[int] = None
    po_number: Optional[str] = None

    class Config:
        from_attributes = True


class StockHistoryRecord(BaseModel):
    """Unified stock-in / stock-out event from stock_entries and stock_exits."""

    id: str
    source_table: str  # stock_entries | stock_exits
    source_id: int
    action: str  # IN | OUT
    inventory_id: Optional[int] = None
    product_name: Optional[str] = None
    article_number: Optional[str] = None
    category: Optional[str] = None
    quantity: int
    technician: str
    timestamp: datetime
    po_number: Optional[str] = None
    ticket_id: Optional[str] = None
    reference: Optional[str] = None
    notes: Optional[str] = None


class StockIn(BaseModel):
    """Schema for receiving stock"""
    ref: str  # Product reference/barcode
    quantity: int
    poId: Optional[str] = None
    technician: str
    category: Optional[str] = None
    brand: Optional[str] = None
    productName: Optional[str] = None
    articleNumber: Optional[str] = None
    serialNumbers: Optional[List[str]] = None
    notes: Optional[str] = None


class StockOut(BaseModel):
    """Schema for assigning stock"""
    productId: str
    quantity: int
    ticketId: str
    technician: str
    notes: Optional[str] = None


class TicketBase(BaseModel):
    """Base ticket schema"""
    title: str
    description: Optional[str] = None
    priority: str = "Medium"
    category: Optional[str] = None
    product_needed: Optional[str] = None


class TicketCreate(TicketBase):
    """Schema for creating a ticket"""
    requester: str


class Ticket(TicketBase):
    """Schema for ticket response"""
    id: str
    status: str
    requester: Optional[str] = None
    assignee: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: mobile/api/inventory.ts
// Inventory API — reads from the inventory table (current on-hand stock)
import api from './axios';

export interface InventoryItem {
  id: number;
  productId: number;
  category?: string;
  articleNumber: string;
  serialNumber?: string;
  quantityAvailable: number;
  trackingType: 'SERIALIZED' | 'BULK';
  status: string;
  receivedBy: string;
  receivedAt?: string;
  purchaseOrderId?: number;
  poNumber?: string;
  // Optional denormalized fields (may be populated by higher-level API wrappers)
  productName?: string;
  brand?: string;
  image?: string;
}

interface InventoryItemResponse {
  id: number;
  product_id: number;
  category?: string;
  article_number: string;
  serial_number?: string;
  quantity_available: number;
  tracking_type: string;
  status: string;
  received_by: string;
  received_at?: string;
  purchase_order_id?: number;
  po_number?: string;
}

function mapInventoryItem(row: InventoryItemResponse): InventoryItem {
  return {
    id: row.id,
    productId: row.product_id,
    category: row.category,
    articleNumber: row.article_number,
    serialNumber: row.serial_number,
    quantityAvailable: row.quantity_available,
    trackingType: row.tracking_type as InventoryItem['trackingType'],
    status: row.status,
    receivedBy: row.received_by,
    receivedAt: row.received_at,
    purchaseOrderId: row.purchase_order_id,
    poNumber: row.po_number,
  };
}

export const inventoryApi = {
  getAll: async (params?: {
    category?: string;
    brand?: string;
    search?: string;
    limit?: number;
  }): Promise<InventoryItem[]> => {
    const { data } = await api.get<InventoryItemResponse[]>('/inventory', { params });
    return data.map(mapInventoryItem);
  },

  getById: async (id: number): Promise<InventoryItem> => {
    const { data } = await api.get<InventoryItemResponse>(`/inventory/${id}`);
    return mapInventoryItem(data);
  },
};
